Makefile

```diff
@@ -7,12 +7,29 @@ DHRY-CFLAGS += -fno-builtin-printf -fno-common -falign-functions=4
 #DHRY-CFLAGS += -DDHRY_ITERS=20000000
 
 SRC = dhry_1.c dhry_2.c strcmp.S
+OBJS = dhry_1.o dhry_2.o strcmp.o
 HDR = dhry.h
 
+ifeq ($(findstring rv32,$(RISCV_ARCH)),rv32)
+RISCV_LD_EMULATION = elf32lriscv
+else
+RISCV_LD_EMULATION = elf64lriscv
+endif
+
 override CFLAGS += $(DHRY-CFLAGS) $(XCFLAGS) -Xlinker --defsym=__stack_size=0x800 -Xlinker --defsym=__heap_size=0x1000
-dhrystone: $(SRC) $(HDR)
-	$(CC) $(CFLAGS) $(SRC) $(LDFLAGS) $(LOADLIBES) $(LDLIBS) -o $@
+
+# Perform an incremental link on object files to move .text sections into .itim so that benchmark code
+# gets placed in Instruction Tightly-Integrated Memory (ITIM) if one exists.
+%.o: $.o.unrenamed move-text-to-itim.lds
+	$(LD) -i -m $(RISCV_LD_EMULATION) -T move-text-to-itim.lds $< -o $@
+
+%.o.unrenamed: %.S $(HDR)
+%.o.unrenamed: %.c $(HDR)
+	$(CC) $(CFLAGS) -c $< -o $@.unrenamed
+
+dhrystone: $(OBJS)
```

Comment:
Does GCC do cross-module optimization when given all of the source together? If so, we don't want to mess with that.
Would it be possible to hack the linker script to move all of the Proc_ and Func_ sections into itim? And strcmp too!

Reply:
Ok, so a bit of experimentation has yielded this result:
```
$ objdump -x release/*.elf | grep itim
  3 .itim         00000a7a  80000000  20400230  00007000  2**3
80000000 l    d  .itim	00000000 .itim
80000028 g     F .itim	00000022 Proc_3
80000000 g       .itim	00000000 metal_segment_itim_target_start
8000082c g     F .itim	00000008 Proc_7
80000140 g     F .itim	0000002c Proc_4
80000838 g     F .itim	00000046 Proc_8
80000880 g     F .itim	0000001c Func_1
800008e4 g     F .itim	00000008 Func_3
8000004c g     F .itim	000000f0 Proc_1
80000188 g     F .itim	000006a2 main
80000a7a g       .itim	00000000 metal_segment_itim_target_end
8000095c g     F .itim	0000011e strcmp
20400230 g       *ABS*	00000000 metal_segment_itim_source_start
80000000 g     F .itim	00000024 Proc_2
800008a0 g     F .itim	00000040 Func_2
800008f0 g     F .itim	0000006a Proc_6
80000170 g     F .itim	00000016 Proc_5
```
that was done by fixing picolibc to put strcmp.S into .text.strcmp, then hacking the linker script a bit. This seems like a generally useful thing to do, and should yield better performance than just placing dhry_*.o in itim as dhrystone depends on strcmp a lot.

Here's my hacked up linker script (this one is for qemu-sifive-e31). One trick was to move .itim before .text so that
sections matching both got dumped into .itim instead of .text

[metal.ramrodata.lds.txt](https://github.com/sifive/benchmark-dhrystone/files/4507002/metal.ramrodata.lds.txt)

Reply:
Yeah, I suppose as long as the ramrodata linker script is doing benchmark-specific things, we could just do this there.

Reply:
ramrodata.lds file is for both coremark and dhrystone, (+ more in future?). So adding specific func signature or file name to the ramrodata.lds would not work. Have a relocation in each specific benchmark Makefile would gives finer control, plus a common ramrodata.lds
We can provide Application Note, if customer want more optimized version.